Signal/src/contact/SystemContactsFetcher.swift

```diff
@@ -417,25 +419,55 @@ class SystemContactsFetcher: NSObject {
         }
     }
 
-    public func fetchIfAlreadyAuthorized(ignoreDebounce: Bool = false) {
+    public func fetchIfAlreadyAuthorized(alwaysNotify: Bool = false) {
         AssertIsOnMainThread()
         guard authorizationStatus == .authorized else {
             return
         }
 
-        updateContacts(completion: nil, ignoreDebounce:ignoreDebounce)
+        updateContacts(completion: nil, alwaysNotify:alwaysNotify)
     }
 
-    private func updateContacts(completion: ((Error?) -> Void)?, ignoreDebounce: Bool = false) {
+    private func tryToAcquireContactFetchLock() -> Bool {
+        var didAcquireLock = false
+        objc_sync_enter(self)
+        if !self.isFetchingContacts {
+            self.isFetchingContacts = true
+            didAcquireLock = true
+        }
+        objc_sync_exit(self)
+        return didAcquireLock
+    }
+
+    private func releaseContactFetchLock() {
+        objc_sync_enter(self)
+        self.isFetchingContacts = false
+        objc_sync_exit(self)
+    }
+
+    private func updateContacts(completion: ((Error?) -> Void)?, alwaysNotify: Bool = false, shouldDebounce: Bool = true) {
         AssertIsOnMainThread()
 
         systemContactsHaveBeenRequestedAtLeastOnce = true
         setupObservationIfNecessary()
 
         DispatchQueue.global().async {
 
-            var fetchedContacts: [Contact]?
+            if shouldDebounce {
+                guard self.tryToAcquireContactFetchLock() else {
+                    Logger.info("\(self.TAG) ignoring redundant system contacts fetch.")
+                    return
+                }
+            }
+            defer {
```

Comment:
Can you put the defer in the previous `if shouldDebounce` block to put the "teardown" code as logically close to the "setup" code as possible.

Reply:
No, because then the `defer` executes when you leave the condition block.

Reply:
I am surprised and appalled.